Mono.Nat/SocketGroup.cs

```diff
@@ -34,9 +34,11 @@ public async Task<(IPAddress, UdpReceiveResult)> ReceiveAsync (CancellationToken
                     try {
                         if (keypair.Key.Available > 0) {
                             var localAddress = ((IPEndPoint) keypair.Key.Client.LocalEndPoint).Address;
-                            var data = await keypair.Key.ReceiveAsync ();
+                            var data = await keypair.Key.ReceiveAsync ().WithCancellation (token);
```

Comment:
This is pretty unsafe. It adds an early exit to this method without cancelling the ReceiveAsync. As a such it opens up the possibility of calling `ReceiveAsync` twice on the same object.

Is this an actual problem you need solved, or did you add this because you felt that it would be cleaner to have cancellation in this method? If this is an actual problem in a real world scenario then we'll have to approach this in a more complex way.

Reply:
You're right of course.  The exception that would be thrown outside of that loop doesn't guarantee nothing will try to read from these sockets again.  I'm not sure how someone is supposed to actually cancel the `ReceiveAsync` short of disposing the `UdpClient` (or whatever that thing was) because the framework offered no such function.  In this case, the `.WithCancellation (token)` probably served no purpose anyway because just before that, the code already confirmed some data was available to read.